### PR TITLE
Adding override code to create RestAPI Lambda Authorizers

### PR DESCRIPTION
**Pull request by sushpatg**
#### Description of changes:

Currently CLI/Rest API override doc is containing override code to create Cognito Authorizer. Submitting pull request adding override code to create and attach lambda authorizers to Rest API path

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [x] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/pages/cli/restapi/override.mdx [hunk 1]
```diff
@@ -152,3 +152,132 @@ user who is invoking the
 endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
 the identity of the user for accessing other services like AWS AppSync or S3.
 
+## Authorize API requests with Lambda Authorizer
+
+In a similar way like Cognito Authorizer, Amplify CLI also does not support Lambda authorizers out-of-box. To implement this functionality, you must override your REST API and add a Lambda authorizer yourself by adding the following code into the `override(...)` function. The following steps are walk-through to create Token-Based Lambda Authorizer.
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:157]:
```suggestion
While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality my overriding your REST API resources. The following steps are walk-through to create token-based Lambda authorizer.
```

File: src/pages/cli/restapi/override.mdx [hunk 2]
```diff
@@ -152,3 +152,132 @@ user who is invoking the
...
+
+First, assuming the Lambda Authorizer function you would like to use as an authorizer is configured with your Amplify Project, create a parameter that resolves to Lambda Function ARN:
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:159]: I'd rewrite this to also link to the flow where you run `amplify add function` and request customers to review there.

Comment (renebrandel) [reply src/pages/cli/restapi/override.mdx:159]: Can you illustrate where I create this parameter? For example in xyz file... or "after you run...".

File: src/pages/cli/restapi/override.mdx [hunk 3]
```diff
@@ -152,3 +152,132 @@ user who is invoking the
...
+    // Use your new Lambda authorizer for security
+    resources.restApi.addPropertyOverride(
+      `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
+      [ { Lambda: [], }, ]
+    );
+  }
+```
+
+
+You can also configure Request Based Lambda Authorizer referring the OpenAPI extension, [`x-amazon-apigateway-authorizer`](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html).
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:278]:
```suggestion
You can also configure request-based Lambda authorizers referenced in the OpenAPI extension, [`x-amazon-apigateway-authorizer`](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html).
```

File: src/pages/cli/restapi/override.mdx [hunk 4]
```diff
@@ -152,3 +152,132 @@ user who is invoking the
...
+Additionally you can add more advanced logic to only use the Lambda authorizer
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:279]: I'd suggest adding a code sample here or exclude it for readability.

File: src/pages/cli/restapi/override.mdx [hunk 5]
```diff
@@ -152,3 +152,132 @@ user who is invoking the
...
+with some paths or methods.
+
+
+When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda Authorizer Function.
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:283]: Show a code snippet here with for the client-side execution

File: src/pages/cli/restapi/override.mdx (hunk 1 above)
Comment (sushpatg) [reply src/pages/cli/restapi/override.mdx]: Hello Rene,
I have updated introduction as suggested.

File: src/pages/cli/restapi/override.mdx (hunk 2 above)
Comment (sushpatg) [reply src/pages/cli/restapi/override.mdx]: Added, steps to create lambda function and mentioned file name to add the code

File: src/pages/cli/restapi/override.mdx (hunk 5 above)
Comment (sushpatg) [reply src/pages/cli/restapi/override.mdx]: Hello,
Referring the documentation for cognito authorizer, I could see client code is mentioned in lib section. Hence added the client code for lambda authorizer in lib section -> Define authorization rules.

File: src/fragments/lib/restapi/js/authz.mdx [hunk 6]
```diff
@@ -73,3 +73,24 @@ postData();
 ```
 
 > Note that the header name, in the above example 'Authorization', is dependent on what you choose during your API Gateway configuration.
+
+## Token-based Lambda Authorization
+
+You need to provide Authorization token to authenticate against API Gateway when using token-based Lambda Authorizer
```
Comment (renebrandel) [inline src/fragments/lib/restapi/js/authz.mdx:79]:
```suggestion
You need to provide an authorization token to authenticate against API Gateway when using token-based Lambda Authorizer
```

File: src/fragments/lib/restapi/js/authz.mdx [hunk 7]
```diff
@@ -73,3 +73,24 @@ postData();
 ```
 
 > Note that the header name, in the above example 'Authorization', is dependent on what you choose during your API Gateway configuration.
+
+## Token-based Lambda Authorization
```
Comment (renebrandel) [inline src/fragments/lib/restapi/js/authz.mdx:77]:
```suggestion
## Token-based Lambda authorization
```

File: src/pages/cli/restapi/override.mdx [hunk 8]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
 endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
 the identity of the user for accessing other services like AWS AppSync or S3.
 
+## Authorize API requests with Lambda Authorizer
+
+While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps are walk-through to create token-based Lambda authorizer.
+
+Firstly, you need to have an  Lambda Authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:159]:
```suggestion
Firstly, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```

File: src/pages/cli/restapi/override.mdx [hunk 9]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+        authorizerResultTtlInSeconds: 0
+      },
+    },
+  });
+```
+
+As API Gateway needs permission to invoke the Authorizer lambda function, add resource based policy to the function using following code:
+
+```ts
+// Adding Resource Based policy to Lambda Authorizer function
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:214]:
```suggestion
// Adding Resource Based policy to Lambda authorizer function
```

File: src/pages/cli/restapi/override.mdx [hunk 10]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+      [ { Lambda: [], }, ]
+    );
+  }
+```
+<Callout>
+Note that you can add more advanced logic to only use the Lambda authorizer
+with some paths or methods.
+</Callout>
+
+When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda Authorizer Function.
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:284]:
```suggestion
When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda authorizer function. 
```

**Review by renebrandel — APPROVED** [review]

File: src/pages/cli/restapi/override.mdx [hunk 11]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
 endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
 the identity of the user for accessing other services like AWS AppSync or S3.
 
+## Authorize API requests with Lambda authorizer
+
+While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps are walk-through to create token-based Lambda authorizer.
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:157]:
```suggestion
While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.
```

File: src/pages/cli/restapi/override.mdx [hunk 12]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+
+Firstly, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:159]:
```suggestion
First, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your Amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```

File: src/pages/cli/restapi/override.mdx [hunk 13]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+
+//  Adding parameter to your Cloud Formation Template for Authorizer function arn
+  resources.addCfnParameter({
+    type: "String",
+    description: "The ARN of an existing Lambda Function to authorize requests",
+    default: "NONE",
+  },
+  functionArnParameter,
+  { "Fn::GetAtt": [`function${functionResourcename}`, "Outputs.Arn"], }
+);
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:177]:
```suggestion
resources.addCfnParameter(
  {
    type: 'String',
    description: 'The ARN of an existing Lambda Function to authorize requests',
    default: 'NONE',
  },
  functionArnParameter,
  { 'Fn::GetAtt': [`function${functionResourcename}`, 'Outputs.Arn'] }
)
```

nit for formatting

File: src/pages/cli/restapi/override.mdx [hunk 14]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+      "x-amazon-apigateway-authorizer": {
+        type: "token",
+        authorizerUri: 
+        {
+          'Fn::Join': ['', ["arn:aws:apigateway:",{ Ref: 'AWS::Region' },":lambda:path/2015-03-31/functions/",{ Ref: functionArnParameter },"/invocations"]],
+        },
+        authorizerResultTtlInSeconds: 0
+      },
+    },
+  });
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:208]:
```suggestion
 // Create the authorizer using the functionArnParameter parameter defined above
resources.restApi.addPropertyOverride('Body.securityDefinitions', {
  Lambda: {
    type: 'apiKey',
    name: 'Authorization',
    in: 'header',
    'x-amazon-apigateway-authtype': 'oauth2',
    'x-amazon-apigateway-authorizer': {
      type: 'token',
      authorizerUri: {
        'Fn::Join': [
          '',
          [
            'arn:aws:apigateway:',
            { Ref: 'AWS::Region' },
            ':lambda:path/2015-03-31/functions/',
            { Ref: functionArnParameter },
            '/invocations',
          ],
        ],
      },
      authorizerResultTtlInSeconds: 0,
    },
  },
})
```

File: src/pages/cli/restapi/override.mdx [hunk 15]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+                    {
+                      "Ref": "<your-API-resource-name>"
+                    },
+                    "/*/*"
+                  ]
+                ]
+              }
+            }
+    },"LambdaAuthorizerResourceBasedPolicy");
+
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:244]:
```suggestion
// Adding Resource Based policy to Lambda authorizer function
resources.addCfnResource(
  {
    type: 'AWS::Lambda::Permission',
    properties: {
      Action: 'lambda:InvokeFunction',
      FunctionName: { Ref: functionArnParameter },
      Principal: 'apigateway.amazonaws.com',
      SourceArn: {
        'Fn::Join': [
          '',
          [
            'arn:aws:execute-api:',
            {
              Ref: 'AWS::Region',
            },
            ':',
            {
              Ref: 'AWS::AccountId',
            },
            ':',
            {
              Ref: '<your-API-resource-name>',
            },
            '/*/*',
          ],
        ],
      },
    },
  },
  'LambdaAuthorizerResourceBasedPolicy'
)
```

File: src/pages/cli/restapi/override.mdx [hunk 16]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+          type: "string",
+        },
+      ]
+    );
+    // Use your new Lambda authorizer for security
+    resources.restApi.addPropertyOverride(
+      `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
+      [ { Lambda: [], }, ]
+    );
+  }
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:277]:
```suggestion
for (const path in resources.restApi.body.paths) {
  // Add the Authorization header as a parameter to requests
  resources.restApi.addPropertyOverride(
    `Body.paths.${path}.x-amazon-apigateway-any-method.parameters`,
    [
      ...resources.restApi.body.paths[path]['x-amazon-apigateway-any-method']
        .parameters,
      {
        name: 'Authorization',
        in: 'header',
        required: false,
        type: 'string',
      },
    ]
  )
  // Use your new Lambda authorizer for security
  resources.restApi.addPropertyOverride(
    `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
    [{ Lambda: [] }]
  )
}
```

File: src/pages/cli/restapi/override.mdx [hunk 17]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+```
+<Callout>
+Note that you can add more advanced logic to only use the Lambda authorizer
+with some paths or methods.
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:281]:
```suggestion
Note that you can add more advanced logic to only use the Lambda authorizer with some paths or methods.
```

File: src/pages/cli/restapi/override.mdx [hunk 18]
```diff
@@ -152,3 +152,133 @@ user who is invoking the
...
+              }
+            }
+    },"LambdaAuthorizerResourceBasedPolicy");
+
+```
+<Callout warning>
+
+Make sure to replace `<your-api-resource-name>` with the name of your Rest API resource.
+This is the name of the folder in `amplify/backend/api` that was created when
+you added an Rest API resource to your Amplify project.
```
Comment (josefaidt) [inline src/pages/cli/restapi/override.mdx:250]:
```suggestion
Make sure to replace `<your-api-resource-name>` with the name of your Rest API resource. This is the name of the folder in `amplify/backend/api` that was created when you added an Rest API resource to your Amplify project.
```

**Review by josefaidt — CHANGES_REQUESTED** [review]
Hey @sushpatg :wave: thank you for taking the time to file this! I left a few comments around formatting and some minor language tweaks, but overall this is really great! Thank you, again!

**Review by josefaidt — DISMISSED** [review]

**Review by josefaidt — APPROVED** [review]

File: src/pages/cli/restapi/override.mdx [hunk 19]
```diff
@@ -152,3 +152,140 @@ user who is invoking the
 endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
 the identity of the user for accessing other services like AWS AppSync or S3.
 
+## Authorize API requests with Lambda authorizer
+
+While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.
```
Comment (renebrandel) [inline src/pages/cli/restapi/override.mdx:157]:
```suggestion
While Amplify CLI does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.
```

**Review by renebrandel — APPROVED** [review]

**Review by josefaidt — APPROVED** [review]

**Review by alharris-at — APPROVED** [review]

**Review by katiegoines — APPROVED** [review]

**Pull request merged**